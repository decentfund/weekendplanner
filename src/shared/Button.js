import React from 'react';
import PropTypes from 'prop-types';

const Button = props => (
  <button className="pure-button" {...props}>
    {props.children}
  </button>
);

Button.defaultProps = {
  children: null,
};

Button.propTypes = {
  children: PropTypes.node,
};

export default Button;
