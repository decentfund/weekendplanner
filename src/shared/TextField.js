import React from 'react';
import PropTypes from 'prop-types';

const TextField = props => <input {...props} />;

export default TextField;
