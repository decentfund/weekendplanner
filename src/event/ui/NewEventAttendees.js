import React from 'react';
import uuidv4 from 'uuid/v4';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextField from '../../shared/TextField';
import Button from '../../shared/Button';
import { addAttendeeRequest, removeAttendeeRequest } from '../new';

export const NewEventAttendees = ({
  accounts,
  currentUserAddress,
  onRemoveClick,
  onAddClick,
}) => (
  <div className="pure-g-1-1">
    {Object.keys(accounts).map(account => (
      <div>
        <TextField disabled={account === currentUserAddress} value={account} />
        {account !== currentUserAddress ? (
          <button onClick={() => onRemoveClick(account)}>Remove</button>
        ) : null}
      </div>
    ))}
    <Button onClick={onAddClick}>Add more</Button>
  </div>
);

NewEventAttendees.propTypes = {
  accounts: PropTypes.object.isRequired, // eslint-disable-line
  currentUserAddress: PropTypes.string,
  onRemoveClick: PropTypes.func,
  onAddClick: PropTypes.func,
};

NewEventAttendees.defaultProps = {
  currentUserAddress: null,
  onRemoveClick: () => {},
  onAddClick: () => {},
};

const mapStateToProps = state => ({
  accounts: state.new.accounts,
});

export default connect(mapStateToProps, {
  onAddClick: dispatch => dispatch(addAttendeeRequest(uuidv4())),
  onRemoveClick: removeAttendeeRequest,
})(NewEventAttendees);
