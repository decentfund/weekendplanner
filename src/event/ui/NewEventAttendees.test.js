import React from 'react';
import { shallow } from 'enzyme';
import { NewEventAttendees, mapStateToProps } from './NewEventAttendees';
import TextField from '../../shared/TextField';
import Button from '../../shared/Button';

const currentUserAddress = '0xabcde';
const newAttendeeAddress = '0xaaaaaa';
const onlyDefaultAttendee = {
  default: {
    address: currentUserAddress,
  },
};

const allAttendees = {
  ...onlyDefaultAttendee,
  '0xaaaaaa': {
    address: newAttendeeAddress,
  },
};

const setupComponent = props => (
  <NewEventAttendees
    accounts={onlyDefaultAttendee}
    currentUserAddress="0xabcde"
    {...props}
  />
);

it('renders NewEventAttendees properly for current user', () => {
  const component = shallow(setupComponent());
  expect(component).toMatchSnapshot();
});

it('renders NewEventAttendees properly for several accounts', () => {
  const component = shallow(
    setupComponent({
      accounts: allAttendees,
    }),
  );
  expect(component).toMatchSnapshot();
});

it('sets proper props for TextField for current user account', () => {
  const component = shallow(
    setupComponent({
      accounts: allAttendees,
    }),
  );

  const currentUserTextField = component.find(TextField).at(0);
  expect(currentUserTextField.prop('disabled')).toBe(true);
  expect(currentUserTextField.prop('value')).toBe(currentUserAddress);
});

it('sets proper props for TextField for new attendee address', () => {
  const onChange = jest.fn();
  const onRemoveClick = jest.fn();
  const component = shallow(
    setupComponent({
      accounts: allAttendees,
      onChangeAddress: onChange,
      onRemoveClick,
    }),
  );

  const currentUserTextField = component.find(TextField).at(1);
  expect(currentUserTextField.prop('disabled')).toBe(false);
  expect(currentUserTextField.prop('value')).toBe(newAttendeeAddress);
  expect(component.contains('Remove')).toBe(true);

  // Handles textfield onChange properly
  const newValue = '123';
  currentUserTextField.simulate('change', { target: { value: newValue } });
  expect(onChange).toBeCalledWith({ id: newAttendeeAddress, value: newValue });

  // Handles remove button click properly
  const removeButtonNewAttendeeAccount = component.find('button').first();
  removeButtonNewAttendeeAccount.simulate('click');
  expect(onRemoveClick).toBeCalledWith(newAttendeeAddress);
});

it('handles onAddClick properly', () => {
  const onAddClick = jest.fn();
  const component = shallow(setupComponent({ onAddClick }));
  const addButton = component.find(Button).first();
  addButton.simulate('click');

  expect(onAddClick).toHaveBeenCalled();
});

// mapStateToProps
it('handles mapStateToProps properly', () => {
  const state = {
    new: {
      attendees: allAttendees,
      attendeesIDs: ['default', newAttendeeAddress],
    },
  };

  expect(mapStateToProps(state)).toEqual({ accounts: allAttendees });
});

// TODO: mapDispatchToProps
