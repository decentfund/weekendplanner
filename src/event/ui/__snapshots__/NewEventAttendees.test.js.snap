exports[`test renders NewEventAttendees properly for current user 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    accounts={
        Object {
            "default": Object {
              "address": "0xabcde",
            },
          }
    }
    currentUserAddress="0xabcde"
    onAddClick={[Function]}
    onChangeAddress={[Function]}
    onRemoveClick={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <div>
            <TextField
                        disabled={true}
                        onChange={[Function]}
                        value="0xabcde" />
</div>,
        ],
        <Button
          onClick={[Function]}>
          Add more
</Button>,
      ],
      "className": "pure-g-1-1",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "default",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextField
              disabled={true}
              onChange={[Function]}
              value="0xabcde" />,
            null,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": true,
              "onChange": [Function],
              "value": "0xabcde",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Add more",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add more",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <div>
              <TextField
                            disabled={true}
                            onChange={[Function]}
                            value="0xabcde" />
</div>,
          ],
          <Button
            onClick={[Function]}>
            Add more
</Button>,
        ],
        "className": "pure-g-1-1",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "default",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextField
                disabled={true}
                onChange={[Function]}
                value="0xabcde" />,
              null,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": true,
                "onChange": [Function],
                "value": "0xabcde",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            null,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Add more",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add more",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`test renders NewEventAttendees properly for several accounts 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    accounts={
        Object {
            "0xaaaaaa": Object {
              "address": "0xaaaaaa",
            },
            "default": Object {
              "address": "0xabcde",
            },
          }
    }
    currentUserAddress="0xabcde"
    onAddClick={[Function]}
    onChangeAddress={[Function]}
    onRemoveClick={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <div>
            <TextField
                        disabled={true}
                        onChange={[Function]}
                        value="0xabcde" />
</div>,
          <div>
            <TextField
                        disabled={false}
                        onChange={[Function]}
                        value="0xaaaaaa" />
            <button
                        onClick={[Function]}>
                        Remove
            </button>
</div>,
        ],
        <Button
          onClick={[Function]}>
          Add more
</Button>,
      ],
      "className": "pure-g-1-1",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "default",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextField
              disabled={true}
              onChange={[Function]}
              value="0xabcde" />,
            null,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": true,
              "onChange": [Function],
              "value": "0xabcde",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "0xaaaaaa",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextField
              disabled={false}
              onChange={[Function]}
              value="0xaaaaaa" />,
            <button
              onClick={[Function]}>
              Remove
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": false,
              "onChange": [Function],
              "value": "0xaaaaaa",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Remove",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Remove",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Add more",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add more",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <div>
              <TextField
                            disabled={true}
                            onChange={[Function]}
                            value="0xabcde" />
</div>,
            <div>
              <TextField
                            disabled={false}
                            onChange={[Function]}
                            value="0xaaaaaa" />
              <button
                            onClick={[Function]}>
                            Remove
              </button>
</div>,
          ],
          <Button
            onClick={[Function]}>
            Add more
</Button>,
        ],
        "className": "pure-g-1-1",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "default",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextField
                disabled={true}
                onChange={[Function]}
                value="0xabcde" />,
              null,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": true,
                "onChange": [Function],
                "value": "0xabcde",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            null,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "0xaaaaaa",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextField
                disabled={false}
                onChange={[Function]}
                value="0xaaaaaa" />,
              <button
                onClick={[Function]}>
                Remove
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": false,
                "onChange": [Function],
                "value": "0xaaaaaa",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Remove",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Remove",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Add more",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add more",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
